File: src/main/gateway/core/pricing.ts
import { readFile } from 'fs/promises'
import type { CostStats, ProviderName, UsageStats } from '../types'

export interface ModelPrice {
  inputPerMTokens: number
  outputPerMTokens: number
  cacheReadPerMTokens?: number
  cacheWrite5mPerMTokens?: number
  cacheWrite1hPerMTokens?: number
}

/** 网关原生计费单位（如 Kiro credit）→ USD 单价 */
export type CreditPriceMap = Record<ProviderName, number>

const BUILTIN_CREDIT_PRICES: CreditPriceMap = {
  // Kiro 订阅：Pro / Pro+ / Power 都是 0.02/credit；overage 0.04
  // 来源：https://kiro.dev/pricing
  kiro: 0.02
}

// 内置主流模型价格表（USD per 1M tokens）
// 数据来源：
// - https://claude.com/pricing
// - https://platform.openai.com/docs/pricing
// 价格变化时通过 ~/.config/gatewayhub/pricing.json 覆盖即可
const BUILTIN: Record<string, ModelPrice> = {
  // ---- Claude 4.x ----
  'claude-opus-4': {
    inputPerMTokens: 15,
    outputPerMTokens: 75,
    cacheReadPerMTokens: 1.5,
    cacheWrite5mPerMTokens: 18.75,
    cacheWrite1hPerMTokens: 30
  },
  'claude-opus-4-1': {
    inputPerMTokens: 15,
    outputPerMTokens: 75,
    cacheReadPerMTokens: 1.5,
    cacheWrite5mPerMTokens: 18.75,
    cacheWrite1hPerMTokens: 30
  },
  'claude-opus-4-5': {
    inputPerMTokens: 15,
    outputPerMTokens: 75,
    cacheReadPerMTokens: 1.5,
    cacheWrite5mPerMTokens: 18.75,
    cacheWrite1hPerMTokens: 30
  },
  'claude-opus-4-7': {
    inputPerMTokens: 15,
    outputPerMTokens: 75,
    cacheReadPerMTokens: 1.5,
    cacheWrite5mPerMTokens: 18.75,
    cacheWrite1hPerMTokens: 30
  },
  'claude-sonnet-4': {
    inputPerMTokens: 3,
    outputPerMTokens: 15,
    cacheReadPerMTokens: 0.3,
    cacheWrite5mPerMTokens: 3.75,
    cacheWrite1hPerMTokens: 6
  },
  'claude-sonnet-4-5': {
    inputPerMTokens: 3,
    outputPerMTokens: 15,
    cacheReadPerMTokens: 0.3,
    cacheWrite5mPerMTokens: 3.75,
    cacheWrite1hPerMTokens: 6
  },
  'claude-sonnet-4-6': {
    inputPerMTokens: 3,
    outputPerMTokens: 15,
    cacheReadPerMTokens: 0.3,
    cacheWrite5mPerMTokens: 3.75,
    cacheWrite1hPerMTokens: 6
  },
  'claude-haiku-4-5': {
    inputPerMTokens: 1,
    outputPerMTokens: 5,
    cacheReadPerMTokens: 0.1,
    cacheWrite5mPerMTokens: 1.25,
    cacheWrite1hPerMTokens: 2
  },
  // ---- Claude 3.x（旧版本）----
  'claude-3-5-sonnet': {
    inputPerMTokens: 3,
    outputPerMTokens: 15,
    cacheReadPerMTokens: 0.3,
    cacheWrite5mPerMTokens: 3.75
  },
  'claude-3-5-haiku': {
    inputPerMTokens: 0.8,
    outputPerMTokens: 4,
    cacheReadPerMTokens: 0.08,
    cacheWrite5mPerMTokens: 1
  },
  'claude-3-opus': {
    inputPerMTokens: 15,
    outputPerMTokens: 75,
    cacheReadPerMTokens: 1.5,
    cacheWrite5mPerMTokens: 18.75
  },
  // ---- OpenAI ----
  'gpt-5': {
    inputPerMTokens: 1.25,
    outputPerMTokens: 10,
    cacheReadPerMTokens: 0.125
  },
  'gpt-5-mini': {
    inputPerMTokens: 0.25,
    outputPerMTokens: 2,
    cacheReadPerMTokens: 0.025
  },
  'gpt-4.1': {
    inputPerMTokens: 2,
    outputPerMTokens: 8,
    cacheReadPerMTokens: 0.5
  },
  'gpt-4.1-mini': {
    inputPerMTokens: 0.4,
    outputPerMTokens: 1.6,
    cacheReadPerMTokens: 0.1
  },
  'gpt-4o': {
    inputPerMTokens: 2.5,
    outputPerMTokens: 10,
    cacheReadPerMTokens: 1.25
  },
  'gpt-4o-mini': {
    inputPerMTokens: 0.15,
    outputPerMTokens: 0.6,
    cacheReadPerMTokens: 0.075
  },
  'o4-mini': {
    inputPerMTokens: 1.1,
    outputPerMTokens: 4.4,
    cacheReadPerMTokens: 0.275
  },
  o3: {
    inputPerMTokens: 2,
    outputPerMTokens: 8,
    cacheReadPerMTokens: 0.5
  }
}

export const ZERO_COST: CostStats = {
  inputUsd: 0,
  outputUsd: 0,
  cacheReadUsd: 0,
  cacheWriteUsd: 0,
  creditsUsd: 0,
  totalUsd: 0,
  currency: 'USD',
  known: false,
  basis: 'none'
}

export interface PricingTableInit {
  modelOverrides?: Record<string, ModelPrice>
  creditOverrides?: CreditPriceMap
}

export class PricingTable {
  private prices: Record<string, ModelPrice>
  private credits: CreditPriceMap
  private resolveCache = new Map<string, ModelPrice | null>()

  constructor(init: Record<string, ModelPrice> | PricingTableInit = {}) {
    // 兼容旧签名：直接传 modelOverrides 对象
    const opts: PricingTableInit =
      'modelOverrides' in init || 'creditOverrides' in init
        ? (init as PricingTableInit)
        : { modelOverrides: init as Record<string, ModelPrice> }
    this.prices = { ...BUILTIN, ...(opts.modelOverrides ?? {}) }
    this.credits = { ...BUILTIN_CREDIT_PRICES, ...(opts.creditOverrides ?? {}) }
  }

  /** 列出当前价格表（含 builtin + override 合并后的结果），用于前端展示 */
  list(): Record<string, ModelPrice> {
    return { ...this.prices }
  }

  /** 列出当前 credit 单价表 */
  listCreditPrices(): CreditPriceMap {
    return { ...this.credits }
  }

  /** 取某个网关的 credit→USD 单价 */
  creditPrice(provider: ProviderName | undefined): number | undefined {
    if (!provider) return undefined
    return this.credits[provider]
  }

  /**
   * 解析模型 ID 到价格条目。
   * 1. 先精确匹配
   * 2. 失败则按 `-` 分段从右往左去掉日期/版本后缀重试
   *    e.g. claude-sonnet-4-5-20250920 → claude-sonnet-4-5 → claude-sonnet-4 → claude-sonnet
   * 3. 同时尝试剥离段内小数尾（claude-opus-4.7 → claude-opus-4），
   *    覆盖 Kiro 把版本号嵌在最后一段的情况
   * 4. 仍然失败返回 undefined
   */
  resolve(model: string | undefined): ModelPrice | undefined {
    if (!model) return undefined
    const normalized = normalizeModelKey(model)
    const cached = this.resolveCache.get(normalized)
    if (cached !== undefined) return cached ?? undefined

    const tryKey = (key: string): ModelPrice | undefined => this.prices[key]

    // 1. 精确匹配
    const exact = tryKey(normalized)
    if (exact) {
      this.resolveCache.set(normalized, exact)
      return exact
    }

    // 2. 段级回退 + 段内小数尾剥离
    const parts = normalized.split('-')
    for (let i = parts.length; i >= 1; i--) {
      const head = parts.slice(0, i).join('-')
      const candidates = [head, stripTrailingDecimal(head)]
      for (const candidate of candidates) {
        if (!candidate || candidate === normalized) continue
        const found = tryKey(candidate)
        if (found) {
          this.resolveCache.set(normalized, found)
          return found
        }
      }
    }

    this.resolveCache.set(normalized, null)
    return undefined
  }

  /**
   * 给定 model + usage 算出费用。
   * - 如果 provider 是 credit 网关（如 kiro）且 usage.credits 已知 → 按 credit×单价 计费
   * - 否则按 token 单价计费；找不到 model 价格条目返回全零
   */
  compute(
    model: string | undefined,
    usage: UsageStats | undefined,
    provider?: ProviderName
  ): CostStats {
    if (!usage) return { ...ZERO_COST }

    // 优先：按上游 credit 计费（Kiro 走这条路）
    const creditPrice = this.creditPrice(provider)
    if (creditPrice !== undefined && typeof usage.credits === 'number' && usage.credits > 0) {
      const creditsUsd = usage.credits * creditPrice
      return {
        inputUsd: 0,
        outputUsd: 0,
        cacheReadUsd: 0,
        cacheWriteUsd: 0,
        creditsUsd,
        totalUsd: creditsUsd,
        currency: 'USD',
        known: true,
        basis: 'credit'
      }
    }

    // 回退：按 token 单价计费
    const price = this.resolve(model)
    if (!price) return { ...ZERO_COST }

    const inputUsd = ((usage.inputTokens || 0) * price.inputPerMTokens) / 1e6
    const outputUsd = ((usage.outputTokens || 0) * price.outputPerMTokens) / 1e6
    const cacheReadUsd = ((usage.cacheReadTokens || 0) * (price.cacheReadPerMTokens ?? 0)) / 1e6
    const cacheWrite5mUsd =
      ((usage.cacheWrite5mTokens || 0) * (price.cacheWrite5mPerMTokens ?? 0)) / 1e6
    const cacheWrite1hUsd =
      ((usage.cacheWrite1hTokens || 0) * (price.cacheWrite1hPerMTokens ?? 0)) / 1e6
    const cacheWriteUsd = cacheWrite5mUsd + cacheWrite1hUsd

    return {
      inputUsd,
      outputUsd,
      cacheReadUsd,
      cacheWriteUsd,
      creditsUsd: 0,
      totalUsd: inputUsd + outputUsd + cacheReadUsd + cacheWriteUsd,
      currency: 'USD',
      known: true,
      basis: 'token'
    }
  }
}

/** 暴露给其他模块用于规整模型 key（去 provider/ 前缀 + 小写） */
export { normalizeModelKey }

/**
 * 从 ~/.config/gatewayhub/pricing.json 异步加载用户覆盖。
 *
 * 文件结构：
 * ```json
 * {
 *   "claude-opus-4-7": { "inputPerMTokens": 15, "outputPerMTokens": 75 },
 *   "__credits__": { "kiro": 0.04 }   // 网关原生计费单位单价（USD/credit）
 * }
 * ```
 */
export async function loadPricingOverrides(filePath: string): Promise<PricingTableInit> {
  try {
    const raw = await readFile(filePath, 'utf8')
    const parsed = JSON.parse(raw)
    if (!parsed || typeof parsed !== 'object') return {}
    const modelOverrides: Record<string, ModelPrice> = {}
    let creditOverrides: CreditPriceMap | undefined
    for (const [key, value] of Object.entries(parsed)) {
      if (key === '__credits__') {
        if (value && typeof value === 'object' && !Array.isArray(value)) {
          creditOverrides = {}
          for (const [provider, price] of Object.entries(value as Record<string, unknown>)) {
            if (typeof price === 'number' && Number.isFinite(price) && price >= 0) {
              creditOverrides[provider] = price
            }
          }
        }
        continue
      }
      if (!value || typeof value !== 'object') continue
      const v = value as Partial<ModelPrice>
      if (typeof v.inputPerMTokens !== 'number' || typeof v.outputPerMTokens !== 'number') continue
      modelOverrides[normalizeModelKey(key)] = {
        inputPerMTokens: v.inputPerMTokens,
        outputPerMTokens: v.outputPerMTokens,
        cacheReadPerMTokens:
          typeof v.cacheReadPerMTokens === 'number' ? v.cacheReadPerMTokens : undefined,
        cacheWrite5mPerMTokens:
          typeof v.cacheWrite5mPerMTokens === 'number' ? v.cacheWrite5mPerMTokens : undefined,
        cacheWrite1hPerMTokens:
          typeof v.cacheWrite1hPerMTokens === 'number' ? v.cacheWrite1hPerMTokens : undefined
      }
    }
    return { modelOverrides, creditOverrides }
  } catch {
    return {}
  }
}

function normalizeModelKey(model: string): string {
  // 去掉 provider/ 前缀，统一小写
  const slash = model.indexOf('/')
  const trimmed = slash >= 0 ? model.slice(slash + 1) : model
  return trimmed.trim().toLowerCase()
}

/** claude-opus-4.7 → claude-opus-4；如果尾段没有小数则返回空串 */
function stripTrailingDecimal(key: string): string {
  const idx = key.lastIndexOf('-')
  if (idx === -1) return ''
  const last = key.slice(idx + 1)
  const dot = last.indexOf('.')
  if (dot === -1) return ''
  return `${key.slice(0, idx + 1)}${last.slice(0, dot)}`
}

File: crates/gateway-core/src/pricing.rs
//! Token/credit pricing — port of `core/pricing.ts`. Builtin table +
//! `pricing.json` overrides + segment-fallback model resolution.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::types::UsageStats;

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelPrice {
    pub input_per_m_tokens: f64,
    pub output_per_m_tokens: f64,
    #[serde(default)]
    pub cache_read_per_m_tokens: Option<f64>,
    #[serde(default)]
    pub cache_write5m_per_m_tokens: Option<f64>,
    #[serde(default)]
    pub cache_write1h_per_m_tokens: Option<f64>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CostStats {
    pub input_usd: f64,
    pub output_usd: f64,
    pub cache_read_usd: f64,
    pub cache_write_usd: f64,
    pub credits_usd: f64,
    pub total_usd: f64,
    pub currency: &'static str,
    pub known: bool,
    /// 'credit' | 'token' | 'free' | 'none'
    pub basis: &'static str,
}

fn zero_cost() -> CostStats {
    CostStats {
        input_usd: 0.0,
        output_usd: 0.0,
        cache_read_usd: 0.0,
        cache_write_usd: 0.0,
        credits_usd: 0.0,
        total_usd: 0.0,
        currency: "USD",
        known: false,
        basis: "none",
    }
}

fn price(input: f64, output: f64) -> ModelPrice {
    ModelPrice {
        input_per_m_tokens: input,
        output_per_m_tokens: output,
        cache_read_per_m_tokens: None,
        cache_write5m_per_m_tokens: None,
        cache_write1h_per_m_tokens: None,
    }
}

fn price_cache(input: f64, output: f64, read: f64, w5m: f64, w1h: f64) -> ModelPrice {
    ModelPrice {
        input_per_m_tokens: input,
        output_per_m_tokens: output,
        cache_read_per_m_tokens: Some(read),
        cache_write5m_per_m_tokens: Some(w5m),
        cache_write1h_per_m_tokens: Some(w1h),
    }
}

fn builtin() -> HashMap<String, ModelPrice> {
    let mut m = HashMap::new();
    for key in [
        "claude-opus-4",
        "claude-opus-4-1",
        "claude-opus-4-5",
        "claude-opus-4-7",
    ] {
        m.insert(key.to_string(), price_cache(15.0, 75.0, 1.5, 18.75, 30.0));
    }
    for key in ["claude-sonnet-4", "claude-sonnet-4-5", "claude-sonnet-4-6"] {
        m.insert(key.to_string(), price_cache(3.0, 15.0, 0.3, 3.75, 6.0));
    }
    m.insert(
        "claude-haiku-4-5".into(),
        price_cache(1.0, 5.0, 0.1, 1.25, 2.0),
    );
    m.insert(
        "claude-3-5-sonnet".into(),
        ModelPrice {
            cache_write1h_per_m_tokens: None,
            ..price_cache(3.0, 15.0, 0.3, 3.75, 0.0)
        },
    );
    m.insert(
        "claude-3-5-haiku".into(),
        ModelPrice {
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.8, 4.0, 0.08, 1.0, 0.0)
        },
    );
    m.insert(
        "claude-3-opus".into(),
        ModelPrice {
            cache_write1h_per_m_tokens: None,
            ..price_cache(15.0, 75.0, 1.5, 18.75, 0.0)
        },
    );
    for key in [
        "gpt-5",
        "gpt-5-codex",
        "gpt-5.1",
        "gpt-5.1-codex",
        "gpt-5.1-codex-max",
    ] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(1.25, 10.0, 0.125, 0.0, 0.0)
            },
        );
    }
    m.insert("gpt-5-pro".into(), price(15.0, 120.0));
    m.insert(
        "gpt-5-mini".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.25, 2.0, 0.025, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-5.1-codex-mini".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.25, 2.0, 0.025, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-5-nano".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.05, 0.4, 0.005, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-4.1".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(2.0, 8.0, 0.5, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-4.1-mini".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.4, 1.6, 0.1, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-4o".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(2.5, 10.0, 1.25, 0.0, 0.0)
        },
    );
    m.insert(
        "gpt-4o-mini".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.15, 0.6, 0.075, 0.0, 0.0)
        },
    );
    m.insert(
        "o4-mini".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(1.1, 4.4, 0.275, 0.0, 0.0)
        },
    );
    m.insert(
        "o3".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(2.0, 8.0, 0.5, 0.0, 0.0)
        },
    );
    // ---- z.ai GLM (glm-5.x share one rate card; flash line is cheaper) ----
    for key in ["glm-5.1", "glm-5.2", "glm-5.3"] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(1.4, 4.4, 0.26, 0.0, 0.0)
            },
        );
    }
    m.insert(
        "glm-5".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(1.0, 3.2, 0.2, 0.0, 0.0)
        },
    );
    for key in ["glm-5.3-flash", "glm-4.7-flashx", "glm-4.5-flash"] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(0.1, 0.5, 0.01, 0.0, 0.0)
            },
        );
    }
    // ---- DeepSeek direct API ----
    // cache_read = the discounted cache-hit input rate; DeepSeek doesn't
    // meter cache writes separately so write lanes stay unpriced.
    for key in [
        "deepseek-v4-flash",
        "deepseek-v4-flash-official",
        "deepseek-chat",
    ] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(0.14, 0.28, 0.0028, 0.0, 0.0)
            },
        );
    }
    for key in ["deepseek-v4-pro", "deepseek-reasoner"] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(0.435, 0.87, 0.0036, 0.0, 0.0)
            },
        );
    }
    // ---- Moonshot Kimi ----
    for key in ["kimi-k2", "kimi-k2.6", "kimi-k2.7-code"] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(0.95, 4.0, 0.16, 0.0, 0.0)
            },
        );
    }
    m.insert(
        "kimi-k3".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(3.0, 15.0, 0.3, 0.0, 0.0)
        },
    );
    // ---- Google Gemini ----
    m.insert(
        "gemini-3-pro".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(2.0, 12.0, 0.2, 0.0, 0.0)
        },
    );
    for key in ["gemini-3-flash", "gemini-2.5-flash"] {
        m.insert(
            key.to_string(),
            ModelPrice {
                cache_write5m_per_m_tokens: None,
                cache_write1h_per_m_tokens: None,
                ..price_cache(0.3, 2.5, 0.03, 0.0, 0.0)
            },
        );
    }
    m.insert(
        "gemini-2.5-pro".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(1.25, 10.0, 0.125, 0.0, 0.0)
        },
    );
    // ---- Alibaba Qwen / xAI Grok / MiniMax ----
    m.insert("qwen3-max".into(), price(1.2, 6.0));
    m.insert("qwen3-coder".into(), price(0.5, 2.0));
    m.insert(
        "grok-4-fast".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(0.2, 0.5, 0.02, 0.0, 0.0)
        },
    );
    m.insert(
        "grok-4".into(),
        ModelPrice {
            cache_write5m_per_m_tokens: None,
            cache_write1h_per_m_tokens: None,
            ..price_cache(3.0, 15.0, 0.3, 0.0, 0.0)
        },
    );
    for key in ["minimax-m2", "minimax-m2.5"] {
        m.insert(key.to_string(), price(0.3, 1.2));
    }
    m
}

fn builtin_credit_prices() -> HashMap<String, f64> {
    // Kiro subscription: Pro/Pro+/Power 0.02/credit (overage 0.04)
    HashMap::from([("kiro".to_string(), 0.02)])
}

pub struct PricingTable {
    prices: HashMap<String, ModelPrice>,
    credits: HashMap<String, f64>,
}

impl Default for PricingTable {
    fn default() -> Self {
        Self::new(None)
    }
}

impl PricingTable {
    pub fn new(overrides_path: Option<&std::path::Path>) -> Self {
        let mut prices = builtin();
        let mut credits = builtin_credit_prices();
        if let Some(path) = overrides_path {
            apply_overrides(path, &mut prices, &mut credits);
        }
        Self { prices, credits }
    }

    pub fn list(&self) -> &HashMap<String, ModelPrice> {
        &self.prices
    }

    pub fn credit_price(&self, provider: &str) -> Option<f64> {
        self.credits.get(provider).copied()
    }

    /// Segment-fallback resolution — `claude-sonnet-4-5-20250920` →
    /// `claude-sonnet-4-5` → `claude-sonnet-4`, plus decimal-tail strip.
    pub fn resolve(&self, model: &str) -> Option<ModelPrice> {
        let key = normalize_model_key(model);
        if key.is_empty() {
            return None;
        }
        if let Some(p) = self.prices.get(&key) {
            return Some(*p);
        }
        let parts: Vec<&str> = key.split('-').collect();
        for i in (1..=parts.len()).rev() {
            let head = parts[..i].join("-");
            for candidate in [head.clone(), strip_trailing_decimal(&head)] {
                if candidate.is_empty() || candidate == key {
                    continue;
                }
                if let Some(p) = self.prices.get(&candidate) {
                    return Some(*p);
                }
            }
        }
        None
    }

    /// `compute` port — `:free` ids are known-zero (aggregator free tier),
    /// then credit-basis (kiro), else token pricing.
    pub fn compute(&self, model: &str, usage: &UsageStats, provider: Option<&str>) -> CostStats {
        // Must run before `resolve`: `deepseek-v4-flash-0731:free` would
        // otherwise fall back to the paid `deepseek-v4-flash` row.
        if normalize_model_key(model).ends_with(":free") {
            return CostStats {
                known: true,
                basis: "free",
                ..zero_cost()
            };
        }
        if let Some(credit_price) = provider.and_then(|p| self.credit_price(p))
            && let Some(credits) = usage.credits.filter(|c| *c > 0.0)
        {
            let credits_usd = credits * credit_price;
            return CostStats {
                credits_usd,
                total_usd: credits_usd,
                known: true,
                basis: "credit",
                ..zero_cost()
            };
        }
        let Some(price) = self.resolve(model) else {
            return zero_cost();
        };
        let input_usd = (usage.input_tokens as f64 * price.input_per_m_tokens) / 1e6;
        let output_usd = (usage.output_tokens as f64 * price.output_per_m_tokens) / 1e6;
        let cache_read_usd = (usage.cache_read_tokens.unwrap_or(0) as f64
            * price.cache_read_per_m_tokens.unwrap_or(0.0))
            / 1e6;
        let cache_write_usd = (usage.cache_write5m_tokens.unwrap_or(0) as f64
            * price.cache_write5m_per_m_tokens.unwrap_or(0.0)
            + usage.cache_write1h_tokens.unwrap_or(0) as f64
                * price.cache_write1h_per_m_tokens.unwrap_or(0.0))
            / 1e6;
        CostStats {
            input_usd,
            output_usd,
            cache_read_usd,
            cache_write_usd,
            credits_usd: 0.0,
            total_usd: input_usd + output_usd + cache_read_usd + cache_write_usd,
            currency: "USD",
            known: true,
            basis: "token",
        }
    }
}

/// `normalizeModelKey` — strip `provider/` prefix, lowercase.
pub fn normalize_model_key(model: &str) -> String {
    let trimmed = match model.find('/') {
        Some(i) => &model[i + 1..],
        None => model,
    };
    trimmed.trim().to_lowercase()
}

/// `claude-opus-4.7` → `claude-opus-4`; no decimal tail → "".
fn strip_trailing_decimal(key: &str) -> String {
    let Some(idx) = key.rfind('-') else {
        return String::new();
    };
    let last = &key[idx + 1..];
    let Some(dot) = last.find('.') else {
        return String::new();
    };
    format!("{}{}", &key[..idx + 1], &last[..dot])
}

/// `loadPricingOverrides` — `pricing.json` with `__credits__` section.
fn apply_overrides(
    path: &std::path::Path,
    prices: &mut HashMap<String, ModelPrice>,
    credits: &mut HashMap<String, f64>,
) {
    let Ok(raw) = std::fs::read_to_string(path) else {
        return;
    };
    let Ok(parsed) = serde_json::from_str::<Value>(&raw) else {
        return;
    };
    let Some(map) = parsed.as_object() else {
        return;
    };
    for (key, value) in map {
        if key == "__credits__" {
            if let Some(obj) = value.as_object() {
                for (provider, price) in obj {
                    if let Some(p) = price.as_f64().filter(|p| p.is_finite() && *p >= 0.0) {
                        credits.insert(provider.clone(), p);
                    }
                }
            }
            continue;
        }
        let Some(v) = value.as_object() else {
            continue;
        };
        let (Some(input), Some(output)) = (
            v.get("inputPerMTokens").and_then(Value::as_f64),
            v.get("outputPerMTokens").and_then(Value::as_f64),
        ) else {
            continue;
        };
        prices.insert(
            normalize_model_key(key),
            ModelPrice {
                input_per_m_tokens: input,
                output_per_m_tokens: output,
                cache_read_per_m_tokens: v.get("cacheReadPerMTokens").and_then(Value::as_f64),
                cache_write5m_per_m_tokens: v.get("cacheWrite5mPerMTokens").and_then(Value::as_f64),
                cache_write1h_per_m_tokens: v.get("cacheWrite1hPerMTokens").and_then(Value::as_f64),
            },
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_falls_back_segments() {
        let table = PricingTable::default();
        assert!(table.resolve("claude-sonnet-4-5-20250920").is_some());
        assert!(table.resolve("claude-opus-4.7").is_some());
        assert!(table.resolve("openrouter/gpt-5").is_some());
        assert!(table.resolve("totally-unknown-model").is_none());
    }

    #[test]
    fn gateway_models_resolve() {
        let table = PricingTable::default();
        for model in [
            "glm-5.3",
            "glm-5.3-flash",
            "deepseek-v4-flash-official",
            "deepseek-v4-pro",
            "kimi-k2.6",
            "gemini-3-pro",
            "minimax-m2.5",
        ] {
            assert!(table.resolve(model).is_some(), "{model}");
        }
        // A `:free` id must price at zero, never fall back to the paid row.
        let cost = table.compute(
            "deepseek-v4-flash-0731:free",
            &UsageStats {
                input_tokens: 1_000_000,
                output_tokens: 1_000_000,
                ..Default::default()
            },
            Some("openrouter"),
        );
        assert!(cost.known && cost.basis == "free" && cost.total_usd == 0.0);
        // GLM cache reads price at the cached-input rate.
        let cost = table.compute(
            "glm-5.3",
            &UsageStats {
                input_tokens: 1_000_000,
                cache_read_tokens: Some(1_000_000),
                ..Default::default()
            },
            Some("traework"),
        );
        assert!((cost.total_usd - (1.4 + 0.26)).abs() < 1e-9);
    }

    #[test]
    fn compute_credit_beats_tokens() {
        let table = PricingTable::default();
        let usage = UsageStats {
            input_tokens: 1000,
            output_tokens: 500,
            credits: Some(2.5),
            ..Default::default()
        };
        let cost = table.compute("claude-sonnet-4-5", &usage, Some("kiro"));
        assert!(cost.known && cost.basis == "credit");
        assert!((cost.total_usd - 0.05).abs() < 1e-9);
        let cost2 = table.compute("claude-sonnet-4-5", &usage, Some("nvidia"));
        assert_eq!(cost2.basis, "token");
    }
}
